src/Filecoin/Address.cpp

```diff
@@ -16,165 +15,283 @@ namespace TW::Filecoin {
+static bool parseActorID(const std::string& string, uint64_t& actorID) {
+    if (string.length() > 20) {
         return false;
     }
-    Type type = getType(data[0]);
-    if (type == Type::Invalid) {
+    const bool onlyDigits = std::all_of(
+        string.begin(),
+        string.end(),
+        [](char ch) { return '0' <= ch && ch <= '9'; }
+    );
+    if (!onlyDigits) {
         return false;
-    } else if (type == Type::ID) {
-        // Verify varuint encoding
-        if (data.size() > 11) {
-            return false;
-        }
-        if (data.size() == 11 && data[10] > 0x01) {
-            return false;
-        }
-        std::size_t i;
-        for (i = 1; i < data.size(); i++) {
-            if ((data[i] & 0x80) == 0) {
-                break;
-            }
-        }
-        return i == data.size() - 1;
-    } else {
-        return data.size() == (1 + Address::payloadSize(type));
     }
-}
 
-static bool isValidID(const std::string& string) {
-    if (string.length() > 22)
-        return false;
-    for (auto i = 2ul; i < string.length(); i++) {
-        if (string[i] < '0' || string[i] > '9') {
-            return false;
-        }
-    }
     try {
         size_t chars;
-        [[maybe_unused]] uint64_t id = std::stoull(string.substr(2), &chars);
+        actorID = std::stoull(string, &chars);
         return chars > 0;
     } catch (...) {
         return false;
     }
 }
 
-static bool isValidBase32(const std::string& string, Address::Type type) {
-    // Check if valid Base32.
-    uint8_t size = Address::payloadSize(type);
-    Data decoded;
-    if (!Base32::decode(string.substr(2), decoded, BASE32_ALPHABET_FILECOIN)) {
-        return false;
+/// Decodes the given `string` as an ActorID.
+/// Please note `bytes` must not contain any prefixes.
+/// https://github.com/paritytech/unsigned-varint/blob/a3a5b8f2bee1f44270629e96541adf805a53d32c/src/decode.rs#L66-L86
+static bool decodeActorID(const Data& bytes, uint64_t& actorID, std::size_t& remainingPos) {
+    const std::size_t maxBytes = 9;
+
+    actorID = 0;
+    remainingPos = 0;
+    for (remainingPos = 0; remainingPos < bytes.size() && remainingPos <= maxBytes; ++remainingPos) {
+        auto byte = bytes[remainingPos];
+        auto k = byte & SCHAR_MAX;
+        actorID |= k << (remainingPos * 7);
+
+        // Check if last.
+        if ((byte & 0x80) == 0) {
+            if (byte == 0 && remainingPos > 0) {
+                // If last byte is zero, it could have been "more minimally" encoded by dropping that trailing zero.
+                return false;
+            }
+            ++remainingPos;
+            return true;
+        }
     }
 
-    // Check size
-    if (decoded.size() != size + checksumSize) {
-        return false;
+    // Couldn't find the last byte so `(byte & 0x80) == 0`.
+    return false;
+}
+
+static void writeActorID(uint64_t actorID, Data& dest) {
+    while (actorID >= 0x80) {
+        dest.push_back(((uint8_t)actorID) | 0x80);
```

Comment:
emplace_back should always be used instead of push_back

Reply:
Isn't `emplace_back` used to push back an item that will be constructed by calling a constructor with the given arguments?
I suppose it will be overhead because of the additional [std::allocator_traits::construct](https://en.cppreference.com/w/cpp/memory/allocator_traits/construct) call.

Reply:
Now I see
> same remark, if u use emplace_back, static_cast should not be required.